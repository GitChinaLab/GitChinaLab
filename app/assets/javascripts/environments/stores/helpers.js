/**
 * @param {Object} environment
 * @returns {Object}
 */
export const setDeployBoard = (oldEnvironmentState, environment) => {
  let parsedEnvironment = environment;
  if (!environment.isFolder && environment.rollout_status) {
    parsedEnvironment = {
      ...environment,
      hasDeployBoard: true,
      isDeployBoardVisible:
        oldEnvironmentState.isDeployBoardVisible === false
          ? oldEnvironmentState.isDeployBoardVisible
          : true,
      deployBoardData:
        environment.rollout_status.status === 'found' ? environment.rollout_status : {},
      isLoadingDeployBoard: environment.rollout_status.status === 'loading',
      isEmptyDeployBoard: environment.rollout_status.status === 'not_found',
    };
  }
  return parsedEnvironment;
};
