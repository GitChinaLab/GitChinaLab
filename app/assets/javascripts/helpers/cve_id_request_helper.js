export function createCveIdRequestIssueBody(fullPath, iid) {
  return `### Vulnerability Submission

**NOTE:** Only maintainers of GitLab-hosted projects may request a CVE for
a vulnerability within their project.

Project issue: ${fullPath}#${iid}

#### Publishing Schedule

After a CVE request is validated, a CVE identifier will be assigned. On what
schedule should the details of the CVE be published?

* [ ] Publish immediately
* [ ] Wait to publish

<!--
Please fill out the yaml codeblock below
-->

\`\`\`yaml
reporter:
  name: "TODO" # "First Last"
  email: "TODO" # "email@domain.tld"
vulnerability:
  description: "TODO" # "[VULNTYPE] in [COMPONENT] in [VENDOR][PRODUCT] [VERSION] allows [ATTACKER] to [IMPACT] via [VECTOR]"
  cwe: "TODO" # "CWE-22" # Path Traversal
  product:
    gitlab_path: "${fullPath}"
    vendor: "TODO" # "Deluxe Sandwich Maker Company"
    name: "TODO" # "Deluxe Sandwich Maker 2"
    affected_versions:
      - "TODO" # "1.2.3"
      - "TODO" # ">1.3.0, <=1.3.9"
    fixed_versions:
      - "TODO" # "1.2.4"
      - "TODO" # "1.3.10"
  impact: "TODO" # "CVSS v3 string" # https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator
  solution: "TODO" # "Upgrade to version 1.2.4 or 1.3.10"
  credit: "TODO"
  references:
    - "TODO" # "https://some.domain.tld/a/reference"
\`\`\`

CVSS scores can be computed by means of the [NVD CVSS Calculator](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator).

/relate ${fullPath}#${iid}
/label ~"devops::secure" ~"group::vulnerability research" ~"vulnerability research::cve" ~"advisory::queued"
  `;
}
