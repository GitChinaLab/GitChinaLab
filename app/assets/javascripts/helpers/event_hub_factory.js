/**
 * An event hub with a Vue instance like API
 *
 * NOTE: There's an [issue open][4] to eventually remove this when some
 * coupling in our codebase has been fixed.
 *
 * NOTE: This is a derivative work from [mitt][1] v1.2.0 which is licensed by
 * [MIT License][2] Â© [Jason Miller][3]
 *
 * [1]: https://github.com/developit/mitt
 * [2]: https://opensource.org/licenses/MIT
 * [3]: https://jasonformat.com/
 * [4]: https://gitlab.com/gitlab-org/gitlab/-/issues/223864
 */
class EventHub {
  constructor() {
    this.$_all = new Map();
  }

  dispose() {
    this.$_all.clear();
  }

  /**
   * Register an event handler for the given type.
   *
   * @param {string|symbol} type Type of event to listen for
   * @param {Function} handler Function to call in response to given event
   */
  $on(type, handler) {
    const handlers = this.$_all.get(type);
    const added = handlers && handlers.push(handler);

    if (!added) {
      this.$_all.set(type, [handler]);
    }
  }

  /**
   * Remove an event handler or all handlers for the given type.
   *
   * @param {string|symbol} type Type of event to unregister `handler`
   * @param {Function} handler Handler function to remove
   */
  $off(type, handler) {
    const handlers = this.$_all.get(type) || [];

    const newHandlers = handler ? handlers.filter((x) => x !== handler) : [];

    if (newHandlers.length) {
      this.$_all.set(type, newHandlers);
    } else {
      this.$_all.delete(type);
    }
  }

  /**
   * Add an event listener to type but only trigger it once
   *
   * @param {string|symbol} type Type of event to listen for
   * @param {Function} handler Handler function to call in response to event
   */
  $once(type, handler) {
    const wrapHandler = (...args) => {
      this.$off(type, wrapHandler);
      handler(...args);
    };
    this.$on(type, wrapHandler);
  }

  /**
   * Invoke all handlers for the given type.
   *
   * @param {string|symbol} type The event type to invoke
   * @param {Any} [evt] Any value passed to each handler
   */
  $emit(type, ...args) {
    const handlers = this.$_all.get(type) || [];

    handlers.forEach((handler) => {
      handler(...args);
    });
  }
}

/**
 * Return a Vue like event hub
 *
 * - $on
 * - $off
 * - $once
 * - $emit
 *
 * Please note, this was once implemented with `mitt`, but since then has been reverted
 * because of some API issues. https://gitlab.com/gitlab-org/gitlab/-/merge_requests/35074
 *
 * We'd like to shy away from using a full fledged Vue instance from this in the future.
 */
export default () => {
  return new EventHub();
};
