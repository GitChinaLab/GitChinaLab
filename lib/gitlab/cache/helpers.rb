# frozen_string_literal: true

module Gitlab
  module Cache
    module Helpers
      # @return [ActiveSupport::Duration]
      DEFAULT_EXPIRY = 1.day

      # @return [ActiveSupport::Cache::Store]
      def cache
        Rails.cache
      end

      def render_cached(obj_or_collection, with:, cache_context: -> (_) { current_user&.cache_key }, expires_in: Gitlab::Cache::Helpers::DEFAULT_EXPIRY, **presenter_args)
        json =
          if obj_or_collection.is_a?(Enumerable)
            cached_collection(
              obj_or_collection,
              presenter: with,
              presenter_args: presenter_args,
              context: cache_context,
              expires_in: expires_in
            )
          else
            cached_object(
              obj_or_collection,
              presenter: with,
              presenter_args: presenter_args,
              context: cache_context,
              expires_in: expires_in
            )
          end

        render Gitlab::Json::PrecompiledJson.new(json)
      end

      private

      # Optionally uses a `Proc` to add context to a cache key
      #
      # @param object [Object] must respond to #cache_key
      # @param context [Proc] a proc that will be called with the object as an argument, and which should return a
      #                       string or array of strings to be combined into the cache key
      # @return [String]
      def contextual_cache_key(presenter, object, context)
        return object.cache_key if context.nil?

        [presenter.class.name, object.cache_key, context.call(object)].flatten.join(":")
      end

      # Used for fetching or rendering a single object
      #
      # @param object [Object] the object to render
      # @param presenter [Grape::Entity]
      # @param presenter_args [Hash] keyword arguments to be passed to the entity
      # @param context [Proc]
      # @param expires_in [ActiveSupport::Duration, Integer] an expiry time for the cache entry
      # @return [String]
      def cached_object(object, presenter:, presenter_args:, context:, expires_in:)
        cache.fetch(contextual_cache_key(presenter, object, context), expires_in: expires_in) do
          Gitlab::Json.dump(presenter.represent(object, **presenter_args).as_json)
        end
      end

      # Used for fetching or rendering multiple objects
      #
      # @param objects [Enumerable<Object>] the objects to render
      # @param presenter [Grape::Entity]
      # @param presenter_args [Hash] keyword arguments to be passed to the entity
      # @param context [Proc]
      # @param expires_in [ActiveSupport::Duration, Integer] an expiry time for the cache entry
      # @return [Array<String>]
      def cached_collection(collection, presenter:, presenter_args:, context:, expires_in:)
        json = fetch_multi(presenter, collection, context: context, expires_in: expires_in) do |obj|
          Gitlab::Json.dump(presenter.represent(obj, **presenter_args).as_json)
        end

        json.values
      end

      # An adapted version of ActiveSupport::Cache::Store#fetch_multi.
      #
      # The original method only provides the missing key to the block,
      # not the missing object, so we have to create a map of cache keys
      # to the objects to allow us to pass the object to the missing value
      # block.
      #
      # The result is that this is functionally identical to `#fetch`.
      def fetch_multi(presenter, *objs, context:, **kwargs)
        objs.flatten!
        map = multi_key_map(presenter, objs, context: context)

        # TODO: `contextual_cache_key` should be constructed based on the guideline https://docs.gitlab.com/ee/development/redis.html#multi-key-commands.
        Gitlab::Instrumentation::RedisClusterValidator.allow_cross_slot_commands do
          cache.fetch_multi(*map.keys, **kwargs) do |key|
            yield map[key]
          end
        end
      end

      # @param objects [Enumerable<Object>] objects which _must_ respond to `#cache_key`
      # @param context [Proc] a proc that can be called to help generate each cache key
      # @return [Hash]
      def multi_key_map(presenter, objects, context:)
        objects.index_by do |object|
          contextual_cache_key(presenter, object, context)
        end
      end
    end
  end
end
