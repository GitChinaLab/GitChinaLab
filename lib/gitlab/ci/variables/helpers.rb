# frozen_string_literal: true

module Gitlab
  module Ci
    module Variables
      module Helpers
        class << self
          def merge_variables(current_vars, new_vars)
            current_vars = transform_from_yaml_variables(current_vars)
            new_vars = transform_from_yaml_variables(new_vars)

            transform_to_yaml_variables(
              current_vars.merge(new_vars)
            )
          end

          def transform_to_yaml_variables(vars)
            vars.to_h.map do |key, value|
              { key: key.to_s, value: value, public: true }
            end
          end

          def transform_from_yaml_variables(vars)
            return vars.stringify_keys if vars.is_a?(Hash)

            vars.to_a.to_h { |var| [var[:key].to_s, var[:value]] }
          end

          def inherit_yaml_variables(from:, to:, inheritance:)
            merge_variables(apply_inheritance(from, inheritance), to)
          end

          private

          def apply_inheritance(variables, inheritance)
            case inheritance
            when true then variables
            when false then {}
            when Array then variables.select { |var| inheritance.include?(var[:key]) }
            end
          end
        end
      end
    end
  end
end
