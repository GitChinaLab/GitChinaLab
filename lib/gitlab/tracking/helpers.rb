# frozen_string_literal: true

module Gitlab
  module Tracking
    module Helpers
      def dnt_enabled?
        Gitlab::Utils.to_boolean(request.headers['DNT'])
      end

      def trackable_html_request?
        request.format.html? && !dnt_enabled?
      end
    end
  end
end
