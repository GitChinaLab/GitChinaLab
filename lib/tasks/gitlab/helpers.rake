# frozen_string_literal: true

# Prevent StateMachine warnings from outputting during a cron task
StateMachines::Machine.ignore_method_conflicts = true if ENV['CRON']

task :gitlab_environment do
  Rake::Task[:environment].invoke unless ENV['SKIP_RAILS_ENV_IN_RAKE']

  extend SystemCheck::Helpers
end
