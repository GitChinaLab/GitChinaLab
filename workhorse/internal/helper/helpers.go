package helper

import (
	"bytes"
	"errors"
	"io/ioutil"
	"mime"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"syscall"

	"github.com/sebest/xff"
	"gitlab.com/gitlab-org/labkit/log"
	"gitlab.com/gitlab-org/labkit/mask"
)

const NginxResponseBufferHeader = "X-Accel-Buffering"

func logErrorWithFields(r *http.Request, err error, fields log.Fields) {
	if err != nil {
		CaptureRavenError(r, err, fields)
	}

	printError(r, err, fields)
}

func CaptureAndFail(w http.ResponseWriter, r *http.Request, err error, msg string, code int) {
	http.Error(w, msg, code)
	logErrorWithFields(r, err, nil)
}

func Fail500(w http.ResponseWriter, r *http.Request, err error) {
	CaptureAndFail(w, r, err, "Internal server error", http.StatusInternalServerError)
}

func Fail500WithFields(w http.ResponseWriter, r *http.Request, err error, fields log.Fields) {
	http.Error(w, "Internal server error", http.StatusInternalServerError)
	logErrorWithFields(r, err, fields)
}

func RequestEntityTooLarge(w http.ResponseWriter, r *http.Request, err error) {
	CaptureAndFail(w, r, err, "Request Entity Too Large", http.StatusRequestEntityTooLarge)
}

func printError(r *http.Request, err error, fields log.Fields) {
	if r != nil {
		entry := log.WithContextFields(r.Context(), log.Fields{
			"method": r.Method,
			"uri":    mask.URL(r.RequestURI),
		})
		entry.WithFields(fields).WithError(err).Error()
	} else {
		log.WithFields(fields).WithError(err).Error("unknown error")
	}
}

func SetNoCacheHeaders(header http.Header) {
	header.Set("Cache-Control", "no-cache, no-store, max-age=0, must-revalidate")
	header.Set("Pragma", "no-cache")
	header.Set("Expires", "Fri, 01 Jan 1990 00:00:00 GMT")
}

func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(path)
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.WithError(err).WithField("url", s).Fatal("urlMustParse")
	}
	return u
}

func HTTPError(w http.ResponseWriter, r *http.Request, error string, code int) {
	if r.ProtoAtLeast(1, 1) {
		// Force client to disconnect if we render request error
		w.Header().Set("Connection", "close")
	}

	http.Error(w, error, code)
}

func HeaderClone(h http.Header) http.Header {
	h2 := make(http.Header, len(h))
	for k, vv := range h {
		vv2 := make([]string, len(vv))
		copy(vv2, vv)
		h2[k] = vv2
	}
	return h2
}

func CleanUpProcessGroup(cmd *exec.Cmd) {
	if cmd == nil {
		return
	}

	process := cmd.Process
	if process != nil && process.Pid > 0 {
		// Send SIGTERM to the process group of cmd
		syscall.Kill(-process.Pid, syscall.SIGTERM)
	}

	// reap our child process
	cmd.Wait()
}

func ExitStatus(err error) (int, bool) {
	exitError, ok := err.(*exec.ExitError)
	if !ok {
		return 0, false
	}

	waitStatus, ok := exitError.Sys().(syscall.WaitStatus)
	if !ok {
		return 0, false
	}

	return waitStatus.ExitStatus(), true
}

func DisableResponseBuffering(w http.ResponseWriter) {
	w.Header().Set(NginxResponseBufferHeader, "no")
}

func AllowResponseBuffering(w http.ResponseWriter) {
	w.Header().Del(NginxResponseBufferHeader)
}

func FixRemoteAddr(r *http.Request) {
	// Unix domain sockets have a remote addr of @. This will make the
	// xff package lookup the X-Forwarded-For address if available.
	if r.RemoteAddr == "@" {
		r.RemoteAddr = "127.0.0.1:0"
	}
	r.RemoteAddr = xff.GetRemoteAddr(r)
}

func SetForwardedFor(newHeaders *http.Header, originalRequest *http.Request) {
	if clientIP, _, err := net.SplitHostPort(originalRequest.RemoteAddr); err == nil {
		var header string

		// If we aren't the first proxy retain prior
		// X-Forwarded-For information as a comma+space
		// separated list and fold multiple headers into one.
		if prior, ok := originalRequest.Header["X-Forwarded-For"]; ok {
			header = strings.Join(prior, ", ") + ", " + clientIP
		} else {
			header = clientIP
		}
		newHeaders.Set("X-Forwarded-For", header)
	}
}

func IsContentType(expected, actual string) bool {
	parsed, _, err := mime.ParseMediaType(actual)
	return err == nil && parsed == expected
}

func IsApplicationJson(r *http.Request) bool {
	contentType := r.Header.Get("Content-Type")
	return IsContentType("application/json", contentType)
}

func ReadRequestBody(w http.ResponseWriter, r *http.Request, maxBodySize int64) ([]byte, error) {
	limitedBody := http.MaxBytesReader(w, r.Body, maxBodySize)
	defer limitedBody.Close()

	return ioutil.ReadAll(limitedBody)
}

func CloneRequestWithNewBody(r *http.Request, body []byte) *http.Request {
	newReq := *r
	newReq.Body = ioutil.NopCloser(bytes.NewReader(body))
	newReq.Header = HeaderClone(r.Header)
	newReq.ContentLength = int64(len(body))
	return &newReq
}
