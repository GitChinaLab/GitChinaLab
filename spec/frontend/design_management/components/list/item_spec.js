import { GlIcon, GlLoadingIcon, GlIntersectionObserver } from '@gitlab/ui';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import VueRouter from 'vue-router';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import Item from '~/design_management/components/list/item.vue';

const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter();

// Referenced from: gitlab_schema.graphql:DesignVersionEvent
const DESIGN_VERSION_EVENT = {
  CREATION: 'CREATION',
  DELETION: 'DELETION',
  MODIFICATION: 'MODIFICATION',
  NO_CHANGE: 'NONE',
};

describe('Design management list item component', () => {
  let wrapper;
  const imgId = 1;
  const imgFilename = 'test';

  const findDesignEvent = () => wrapper.findByTestId('design-event');
  const findImgFilename = (id = imgId) => wrapper.findByTestId(`design-img-filename-${id}`);
  const findEventIcon = () => findDesignEvent().find(GlIcon);
  const findLoadingIcon = () => wrapper.find(GlLoadingIcon);

  function createComponent({
    notesCount = 0,
    event = DESIGN_VERSION_EVENT.NO_CHANGE,
    isUploading = false,
    imageLoading = false,
  } = {}) {
    wrapper = extendedWrapper(
      shallowMount(Item, {
        localVue,
        router,
        propsData: {
          id: imgId,
          filename: imgFilename,
          image: 'http://via.placeholder.com/300',
          isUploading,
          event,
          notesCount,
          updatedAt: '01-01-2019',
        },
        data() {
          return {
            imageLoading,
          };
        },
        stubs: ['router-link'],
      }),
    );
  }

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when item is not in view', () => {
    it('image is not rendered', () => {
      createComponent();

      const image = wrapper.find('img');
      expect(image.attributes('src')).toBe('');
    });
  });

  describe('when item appears in view', () => {
    let image;
    let glIntersectionObserver;

    beforeEach(() => {
      createComponent();
      image = wrapper.find('img');
      glIntersectionObserver = wrapper.find(GlIntersectionObserver);

      glIntersectionObserver.vm.$emit('appear');
      return wrapper.vm.$nextTick();
    });

    it('renders a tooltip', () => {
      expect(findImgFilename().attributes('title')).toEqual(imgFilename);
    });

    describe('before image is loaded', () => {
      it('renders loading spinner', () => {
        expect(wrapper.find(GlLoadingIcon).exists()).toBe(true);
      });
    });

    describe('after image is loaded', () => {
      beforeEach(() => {
        image.trigger('load');
        return wrapper.vm.$nextTick();
      });

      it('renders an image', () => {
        expect(image.attributes('src')).toBe('http://via.placeholder.com/300');
        expect(image.isVisible()).toBe(true);
      });

      it('renders media broken icon when image onerror triggered', () => {
        image.trigger('error');
        return wrapper.vm.$nextTick().then(() => {
          expect(image.isVisible()).toBe(false);
          expect(wrapper.find(GlIcon).element).toMatchSnapshot();
        });
      });

      describe('when imageV432x230 and image provided', () => {
        it('renders imageV432x230 image', () => {
          const mockSrc = 'mock-imageV432x230-url';
          wrapper.setProps({ imageV432x230: mockSrc });

          return wrapper.vm.$nextTick().then(() => {
            expect(image.attributes('src')).toBe(mockSrc);
          });
        });
      });

      describe('when image disappears from view and then reappears', () => {
        beforeEach(() => {
          glIntersectionObserver.vm.$emit('appear');
          return wrapper.vm.$nextTick();
        });

        it('renders an image', () => {
          expect(image.isVisible()).toBe(true);
        });
      });
    });
  });

  describe('with notes', () => {
    it('renders item with single comment', () => {
      createComponent({ notesCount: 1 });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders item with multiple comments', () => {
      createComponent({ notesCount: 2 });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  it('renders loading spinner when isUploading is true', () => {
    createComponent({ isUploading: true });

    expect(findLoadingIcon().exists()).toBe(true);
  });

  it('renders item with no status icon for none event', () => {
    createComponent();

    expect(findDesignEvent().exists()).toBe(false);
  });

  describe('with associated event', () => {
    it.each`
      event                                | icon                     | className
      ${DESIGN_VERSION_EVENT.MODIFICATION} | ${'file-modified-solid'} | ${'text-primary-500'}
      ${DESIGN_VERSION_EVENT.DELETION}     | ${'file-deletion-solid'} | ${'text-danger-500'}
      ${DESIGN_VERSION_EVENT.CREATION}     | ${'file-addition-solid'} | ${'text-success-500'}
    `('renders item with correct status icon for $event event', ({ event, icon, className }) => {
      createComponent({ event });
      const eventIcon = findEventIcon();

      expect(eventIcon.exists()).toBe(true);
      expect(eventIcon.props('name')).toBe(icon);
      expect(eventIcon.classes()).toContain(className);
    });
  });
});
