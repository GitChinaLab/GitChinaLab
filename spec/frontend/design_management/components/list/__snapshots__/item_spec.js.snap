// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design management list item component when item appears in view after image is loaded renders media broken icon when image onerror triggered 1`] = `
<gl-icon-stub
  class="text-secondary"
  name="media-broken"
  size="32"
/>
`;

exports[`Design management list item component with notes renders item with multiple comments 1`] = `
<router-link-stub
  ariacurrentvalue="page"
  class="card gl-cursor-pointer text-plain js-design-list-item design-list-item design-list-item-new"
  event="click"
  tag="a"
  to="[object Object]"
>
  <div
    class="card-body gl-p-0 gl-display-flex gl-align-items-center gl-justify-content-center gl-overflow-hidden gl-relative"
  >
    <!---->
     
    <gl-intersection-observer-stub>
      <!---->
       
      <img
        alt="test"
        class="gl-display-block gl-mx-auto gl-max-w-full gl-max-h-full design-img"
        data-qa-filename="test"
        data-qa-selector="design_image"
        data-testid="design-img-1"
        src=""
      />
    </gl-intersection-observer-stub>
  </div>
   
  <div
    class="card-footer gl-display-flex gl-w-full"
  >
    <div
      class="gl-display-flex gl-flex-direction-column str-truncated-100"
    >
      <span
        class="gl-font-weight-bold str-truncated-100"
        data-qa-selector="design_file_name"
        data-testid="design-img-filename-1"
        title="test"
      >
        test
      </span>
       
      <span
        class="str-truncated-100"
      >
        
        Updated 
        <timeago-stub
          cssclass=""
          time="01-01-2019"
          tooltipplacement="bottom"
        />
      </span>
    </div>
     
    <div
      class="gl-ml-auto gl-display-flex gl-align-items-center gl-text-gray-500"
    >
      <gl-icon-stub
        class="gl-ml-2"
        name="comments"
        size="16"
      />
       
      <span
        aria-label="2 comments"
        class="gl-ml-2"
      >
        
        2
      
      </span>
    </div>
  </div>
</router-link-stub>
`;

exports[`Design management list item component with notes renders item with single comment 1`] = `
<router-link-stub
  ariacurrentvalue="page"
  class="card gl-cursor-pointer text-plain js-design-list-item design-list-item design-list-item-new"
  event="click"
  tag="a"
  to="[object Object]"
>
  <div
    class="card-body gl-p-0 gl-display-flex gl-align-items-center gl-justify-content-center gl-overflow-hidden gl-relative"
  >
    <!---->
     
    <gl-intersection-observer-stub>
      <!---->
       
      <img
        alt="test"
        class="gl-display-block gl-mx-auto gl-max-w-full gl-max-h-full design-img"
        data-qa-filename="test"
        data-qa-selector="design_image"
        data-testid="design-img-1"
        src=""
      />
    </gl-intersection-observer-stub>
  </div>
   
  <div
    class="card-footer gl-display-flex gl-w-full"
  >
    <div
      class="gl-display-flex gl-flex-direction-column str-truncated-100"
    >
      <span
        class="gl-font-weight-bold str-truncated-100"
        data-qa-selector="design_file_name"
        data-testid="design-img-filename-1"
        title="test"
      >
        test
      </span>
       
      <span
        class="str-truncated-100"
      >
        
        Updated 
        <timeago-stub
          cssclass=""
          time="01-01-2019"
          tooltipplacement="bottom"
        />
      </span>
    </div>
     
    <div
      class="gl-ml-auto gl-display-flex gl-align-items-center gl-text-gray-500"
    >
      <gl-icon-stub
        class="gl-ml-2"
        name="comments"
        size="16"
      />
       
      <span
        aria-label="1 comment"
        class="gl-ml-2"
      >
        
        1
      
      </span>
    </div>
  </div>
</router-link-stub>
`;
